src/main/java/org/vaadin/viritin/ConcurrentSortableLazyList.java

```diff
@@ -0,0 +1,104 @@
+package org.vaadin.viritin;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A concurrent version of the {@link SortableLazyList}.
+ * <p>
+ * This implementation locks calls to {@link #size()}, {@link #findEntities(int)}, {@link #initCacheFormPage(int)}
+ * and {@link #get(int)}, enabling access from multiple threads at the same time.
+ *
+ * @param <T> The type of the objects in the list
+ * @author Jesenko Mehmedbasic
+ */
+public class ConcurrentSortableLazyList<T> extends SortableLazyList<T> {
+    private final Lock lock = new ReentrantLock();
+    private long lockDelay = 10_000;
+
+    public ConcurrentSortableLazyList(SortableEntityProvider<T> dataProvider) {
+        super(dataProvider);
+    }
+
+    public ConcurrentSortableLazyList(SortableEntityProvider<T> dataProvider, int pageSize) {
+        super(dataProvider, pageSize);
+    }
+
+    public ConcurrentSortableLazyList(SortablePagingProvider<T> pageProvider, CountProvider countProvider) {
+        super(pageProvider, countProvider);
+    }
+
+    public ConcurrentSortableLazyList(SortablePagingProvider<T> pageProvider, CountProvider countProvider, int pageSize) {
+        super(pageProvider, countProvider, pageSize);
+    }
+
+    public ConcurrentSortableLazyList(MultiSortablePagingProvider<T> pageProvider, CountProvider countProvider, int pageSize) {
+        super(pageProvider, countProvider, pageSize);
+    }
+
+    /**
+     * Sets the amount of time the list waits for a lock.
+     *
+     * @param time the time in milliseconds.
+     */
+    public void setLockDelay(long time) {
+        if (time < 0) {
+            throw new IllegalArgumentException("The time argument must be greater than zero.");
+        }
+        this.lockDelay = time;
+    }
+
+    @Override
+    public int size() {
+        try {
+            tryLock();
```

Comment:
Wouldn’t it be better to move `tryLock()` out of this try-block? Otherwise the finally might try to unlock a non-existing lock and throw (another) exception.